ligo/src/token/fa2.mligo

```diff
@@ -0,0 +1,176 @@
+// SPDX-FileCopyrightText: 2020 TQ Tezos
+// SPDX-License-Identifier: LicenseRef-MIT-TQ
+
+// Corresponds to Token/FA2.hs module
+
+#include "../types.mligo"
+
+let frozen_token_id: nat = 0n
+let unfrozen_token_id: nat = 1n
+// let (|>) x f = f x
+
+// -----------------------------------------------------------------
+// Helper
+// -----------------------------------------------------------------
+
+let ensure_not_migrated (storage : storage): bool =
+  match storage.migration_status with
+      Not_in_migration -> true
+    | MigratingTo (p) -> true
+    | MigratedTo (p) -> (failwith ("MIGRATED") : bool)
```

Comment:
Yeah, I was planning to use `unit` type as well (It supposed to be `unit` but I forgot to update, and it applies to all various failwith functions).

I did it this way since it supposed to resembles a guard check, for example in stablecoin pascaligo (different function but serve similar purpose):

```
function ensure_not_paused
  ( const store : storage
  ) : unit is
  fail_on
    ( store.paused
    , "CONTRACT_PAUSED"
    )

function transfer
  ( const param: transfer_param
  ; const store: storage
  ; const full_param: closed_parameter
  ) : entrypoint is block
{ ensure_not_paused(store)
....
}
```

In this usage, we don't really care about the return type.

I'm trying to achieve something similar in cameligo, but seems not possible, so that's why I use various `cond_` which means we don't care the about the return value.

So what do you think about this, Pasquale? Should I update as you suggested earlier or keep it as unit.

Reply:
I still think we should use the `storage` version.

First of all (not that important) because it seems more in `cameLIGO`'s style to only return useful types, given that it does not have a clean way to ignore the result of calling the function.

Secondly, and more importantly, because it's more efficient: with the `bool` or `unit` version there is a `PUSH`/`UNIT` instruction that does not get optimized away and so increases the cost (however little that may be).
With the `storage` version it doesn't happen because the return is already just below in the stack.
Copied from the resulting code, it's:
```
IF_LEFT { DROP ; UNIT } { IF_LEFT { DROP ; UNIT } { DROP ; PUSH string "MIGRATED" ; FAILWITH } } } ;
```
vs
```
IF_LEFT { DROP } { IF_LEFT { DROP } { DROP 2 ; PUSH string "MIGRATED" ; FAILWITH } } } ;
```

Reply:
I have two more requests about this function:

1. define it as `[@inline]`. Just as the `ensure_not_paused` your mentioned is, this will most likely be more efficient.
2. move it to `management.mligo` and encase it between this:
```
#if !MANAGEMENT_H
#define MANAGEMENT_H

<definition here>

#endif  // MANAGEMENT_H included
```
there. For an explanation why see the note about shared functions here: #100

Reply:
Currently only one function from the entire module is enclosed into `MANAGEMENT_H`, probably that's not what you meant? Like, it looks like it's worth deduplicating either the entire module, or nothing.

Reply:
Yeah that's what I meant because this should be left to #100 to solve (and there are no problems now that the endpoint functions are not defined), there I more clearly specified that we need to end up with a complete solution:
> We can either use the existing source file and put all definitions in the block or make a separate one just for the 2 checking functions instead. Your pick.

Reply:
Btw, I also started writing checks so that they return `unit`, and I observe that performance aspect of this seems to be not so critical to me.

Note that we want to return `unit` only for checks, and checks are normally small which means that we always want to inline them (if we don't do this, then checks get extracted into a lambda annotated with storage/config type which is enormously huge, takes about 250 bytes of op size).

And for inlined functions returning `unit` does not add unnecessary `UNIT; DROP`.

Whether is returning `unit` comply with the CameLIGO style (in case if they have a reasonable style with respect to their featureset :thinking:) is also an insteresting question. For instance, if I return non-unit in a check, I have to annotate `failwith` with the returned type, so both approaches seem to be in some sort unpleasant.

Also there is a point about that distinguishing read-only and state-updating methods in their signatures is good, though I don't know whether will we encounter cases where this matters.

So probably I will just follow the style we came up with in here, but if points above make sense to you then next time we can keep in mind that both approaches make sense. Or decide that picking cameLIGO was a mistake :smile:

Reply:
I'm fairly sure I had checked the function as inlined too and it did have an additional `UNIT` instruction in it, but probably it was because it contained a `match` in cameLIGO and it wasn't able to optimize that away.

The improvement is very small either way, it only saves _maybe_ a `DROP` and some `UNIT`s.

I agree on the rest, after this is done we should consider weather we can make better/different choices next.